clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp

```diff
@@ -110,15 +117,32 @@ bool UseAfterMoveFinder::find(Stmt *CodeBlock, const Expr *MovingCall,
   BlockMap = std::make_unique<StmtToBlockMap>(TheCFG.get(), Context);
   Visited.clear();
 
-  const CFGBlock *Block = BlockMap->blockContainingStmt(MovingCall);
-  if (!Block) {
+  const CFGBlock *MoveBlock = BlockMap->blockContainingStmt(MovingCall);
+  if (!MoveBlock) {
     // This can happen if MovingCall is in a constructor initializer, which is
     // not included in the CFG because the CFG is built only from the function
     // body.
-    Block = &TheCFG->getEntry();
+    MoveBlock = &TheCFG->getEntry();
   }
 
-  return findInternal(Block, MovingCall, MovedVariable, TheUseAfterMove);
+  bool Found = findInternal(MoveBlock, MovingCall, MovedVariable->getDecl(),
+                            TheUseAfterMove);
+
+  if (Found) {
+    if (const CFGBlock *UseBlock =
+            BlockMap->blockContainingStmt(TheUseAfterMove->DeclRef)) {
+      // Does the use happen in a later loop iteration than the move?
+      // - If they are in the same CFG block, we know the use happened in a
+      //   later iteration if we visited that block a second time.
+      // - Otherwise, we know the use happened in a later iteration if the
+      //   move is reachable from the use.
+      auto CFA = std::make_unique<CFGReverseBlockReachabilityAnalysis>(*TheCFG);
+      TheUseAfterMove->UseHappensInLaterLoopIteration =
+          UseBlock == MoveBlock ? Visited.contains(UseBlock)
+                                : CFA->isReachable(UseBlock, MoveBlock);
```

Comment:
`CFA` does not need to be a unique_ptr, you can simply create a `CFGReverseBlockReachabilityAnalysis` on the stack.

Reply:
`CFGReverseBlockReachabilityAnalysis` comes with two member variables of `llvm::BitVector` and `llvm::DenseMap<>`, which could be potentially too large to put on the stack, i thought. that's why i allocated it on heap. but if you believe it's safe to do so. will allocate it on stack.

Reply:
Drive-by comment: These types are fine to put on the stack. Note that the elements contained in these containers are stored on the heap in any case. For example, in the case of `DenseMap`, this storage is allocated by `DenseMap::allocateBuckets()`. The only things contained in `DenseMap` itself (and hence stored on the stack) are a pointer to the bucket array `Buckets` and some counters (`NumEntries`, `NumBuckets`, `NumTombstones`).